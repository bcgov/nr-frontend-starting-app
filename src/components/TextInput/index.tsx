import React, { ChangeEventHandler, KeyboardEventHandler } from 'react';

interface InputProps {
  id: string,
  isValid?: boolean,
  size?: string,
  feedback?: string,
  onChangeHandler?: ChangeEventHandler<HTMLInputElement>,
  inputValue?: string,
  onKeyDownHandler?: KeyboardEventHandler<HTMLInputElement>,
  disabled?: boolean,
  readonly?: boolean
}

const TextInput = ({
  id,
  isValid,
  size = 'form-control',
  feedback = '',
  onChangeHandler,
  inputValue = '',
  onKeyDownHandler,
  disabled = false,
  readonly = false
}: InputProps) => {
  let valid = '';

  if (typeof isValid === 'boolean') {
    valid = isValid ? 'is-valid' : 'is-invalid';
  }

  const inputClasses = `${size} ${valid}`;
  const divClassName = isValid ? 'valid-feedback' : 'invalid-feedback';

  return (
    <>
      <input
        data-testid="input-element"
        type="text"
        id={id}
        className={inputClasses}
        onChange={onChangeHandler}
        value={inputValue}
        onKeyDown={onKeyDownHandler}
        disabled={disabled}
        readOnly={readonly}
      />
      <div
        data-testid="feedback-element"
        className={divClassName}
      >
        { feedback }
      </div>
    </>
  );
};

export default TextInput;
