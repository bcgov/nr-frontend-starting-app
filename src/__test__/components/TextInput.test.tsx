/* eslint-disable no-undef */
import { render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import TextInput from '../../components/TextInput';

describe('the TextInput component', () => {
  it('should have no feedback message when none is set', () => {
    const { getByTestId } = render(<TextInput id="test1" />);
    expect(getByTestId('feedback-element').textContent).toBe('');
  });

  it('should have the correct valid feedback message', () => {
    const { getByTestId } = render(<TextInput id="test2" isValid feedback="valid field" />);
    expect(getByTestId('feedback-element').textContent).toBe('valid field');
  });

  it('should have the correct invalid feedback message', () => {
    const { getByTestId } = render(<TextInput id="test3" isValid={false} feedback="invalid field" />);
    expect(getByTestId('feedback-element').textContent).toBe('invalid field');
  });

  it('should match the snapshot for the neutral input', () => {
    const tree = renderer
      .create(<TextInput id="test1" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot for the success input', () => {
    const tree = renderer
      .create(<TextInput id="test2" isValid feedback="valid field" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot for the fail input', () => {
    const tree = renderer
      .create(<TextInput id="test3" isValid={false} feedback="invalid field" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
