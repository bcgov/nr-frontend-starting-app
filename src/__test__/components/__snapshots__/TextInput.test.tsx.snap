// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly fail input 1`] = `
[
  <input
    className="form-control is-invalid"
    id="test3"
    onBlur="undefined"
    type="text"
  />,
  <div
    className="invalid-feedback"
  >
    invalid field
  </div>,
]
`;

exports[`renders correctly neutral input 1`] = `
[
  <input
    className="form-control is-valid"
    id="test1"
    onBlur="undefined"
    type="text"
  />,
  <div
    className="valid-feedback"
  >
    valid field
  </div>,
]
`;

exports[`renders correctly success input 1`] = `
[
  <input
    className="form-control is-valid"
    id="test2"
    onBlur="undefined"
    type="text"
  />,
  <div
    className="valid-feedback"
  >
    valid field
  </div>,
]
`;
